### ISSUE TITLE
Cannot understant the code at README.md of pair-pm

### ISSUE THREAD

**Issue by heyzude** [RLHFlow/RLHF-Reward-Modeling#8]
Hi, I'm very confused by your code.

```
logit_A = output.logits[0, -1, token_id_A].item()
logit_B = output.logits[0, -1, token_id_B].item()
```

I guess the last dimension of the tensor `output.logits` would be same as the vocab_size.
But why `token_id_A`-th and `token_id_B`-th index element matter? (they are 32 and 33 respectively, seemingly random to me).

Also, I found that `avg_prob_chosen` (=`np.mean(probs_chosen)`) is exactly 0.5. How this can be?
Since the positions of two responses change as we iterate through `for chosen_position in [0, 1]:`, the sum of the probs is very unlikely to be 1, but it is...

**Comment by WeiXiongUST** [comment]
Hi, thanks for your interest in the preference model. 

The preference model is used to predict P(A | prompt) and P(B | prompt). The two lines of code compute the logit of decoding A and B, respectively and we only use these two logits. 

Since we iterate through for chosen_position in [0, 1], the preferred response (labelled by the human) appears in A for the first time and B in the second time. Therefore, accordingly, the logit for choosing the correct one is taken as the logit_A (correct answer for the first round d) and logit_B (correct answer for the second round), respectively.
```python
logit_chosen = [logit_A, logit_B][chosen_position]
```

A specific example is as follows. We have two responses, "GOOD" and "BAD" and the first one is labelled as preferred. For the first time, we format the prompt as 

```python
[CONTEXT] \n\n<turn> user\n Some question here... [RESPONSE A] GOOD. [RESPONSE B] BAD.
```
The preference model should decode A because the RESPONSE A should be preferred in this case. In the second round, the prompt is formatted as 
```python
[CONTEXT] \n\n<turn> user\n Some question here... [RESPONSE A] BAD. [RESPONSE B] GOOD.
```
The preference model should decode B because the RESPONSE B should be preferred in this case (so we take the logit_B in this case).

**Comment by heyzude** [comment]
I understood. Thanks!
However, the fact that `avg_prob_chosen == 0.5` is `True` still strikes me.

**Comment by WeiXiongUST** [comment]
This happens more when we quantize the model (the default setting of reward bench) but it should not happen too frequently in practice. 

Also I think I should put the technical report https://arxiv.org/pdf/2405.07863 in the readme where the detailed description of the preference modeling is provided.
